### DISCUSSION TITLE
Need a clarification on tcp-echo-server example

### DISCUSSION THREAD

**Discussion by realyukii** [libuv/libuv#4363] [Q&A] [answered]
I'm sorry if this question sounds like a dumb question, but I'll ask it anyway.

Why isn't the req variable, which is allocated dynamically using malloc, freed within the `echo_read` function? I notice that only `buf->base` is being freed in the function.

Will this function be safe from memory leaks? Is it acceptable to not free variables allocated by malloc?

here's the snippet code that I'm talking about:
```c
void echo_read(uv_stream_t *client, ssize_t nread, const uv_buf_t *buf) {
    if (nread > 0) {
        write_req_t *req = (write_req_t*) malloc(sizeof(write_req_t));
        req->buf = uv_buf_init(buf->base, nread);
        uv_write((uv_write_t*) req, client, &req->buf, 1, echo_write);
        return;
    }
    if (nread < 0) {
        if (nread != UV_EOF)
            fprintf(stderr, "Read error %s\n", uv_err_name(nread));
        uv_close((uv_handle_t*) client, on_close);
    }

    free(buf->base);
}
```

**Comment by bnoordhuis** [comment, marked as answer]
It's freed in echo_write. The uv_write_t needs to persist until that function is called.

**Reply by realyukii** [reply]
thanks for the answers!
By the way, would you be willing to address my [other question](https://github.com/libuv/libuv/issues/3266#issuecomment-1996889588) as well?
